### PR TITLE
Code review time

### PR DESCRIPTION
**Pull request by JaniM**

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+const { Client, GatewayIntentBits, Permissions, PermissionsBitField } = pkg;
+const client = new Client({
+  intents: [
+    GatewayIntentBits.Guilds,
+    GatewayIntentBits.GuildMessages,
+    GatewayIntentBits.MessageContent,
+  ],
+});
+
+const SendCommandsList = () => {
```
Comment (JaniM) [inline bot.js]: Considering this function doesn't send anything, the name is quite confusing.

File: src/live.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+const youtube = google.youtube({
+  version: "v3",
+  auth: process.env.YOUTUBE_API_KEY,
+});
+
+export async function hourlyLiveStreamNotifications(
+  client,
+  scheduledStartTime
+) {
```
Comment (JaniM) [inline src/live.js:17]: This function would be significantly easier to reason about if it didn't update `scheduledStartTime` directly, but instead always returned an array of new schedules.

File: src/live.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      messages.find((message) => {
+        // Check if the message includes the liveStreamVideo URL
+        return message.content.includes(liveStreamVideo.videoURL);
+      })
+    );
+    for (const channel of notificationChannels) {
+      if (!messageStack) {
+        client.channels.cache.get(channel).send(message);
+        console.log(`Posted ${message} to ${channel}`);
+      }
```
Comment (JaniM) [inline src/live.js:77]: Bug? We don't send a notification if it has been sent on *any* notification channel before.

File: bot.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            channel.send(message);
+          } else {
+            console.error(`Channel with ID ${channelId} not found in cache.`);
+          }
+        }
+
+        // Remove the scheduled video from the list
+        const index = scheduledStartTime.indexOf(scheduledVideo);
+        if (index !== -1) {
+          scheduledStartTime.splice(index, 1);
```
Comment (JaniM) [inline bot.js]: This breaks the iterator used by the outer loop.

File: bot.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        }
+      }
+    }
+  }, 10 * 60 * 1000);
+});
+
+client.on("messageCreate", (discordMessageStack) => {
```
Comment (JaniM) [inline bot.js]: Message stack? Isnn't this just a single message.

File: bot.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  messageStack
+) {
+  const doesUserHavePermission = messageStack.member.permissions.has([
+    PermissionsBitField.Flags.KickMembers,
+    PermissionsBitField.Flags.BanMembers,
+  ])
+    ? true
+    : false;
+
+  if (commandSentByUser === "add") {
```
Comment (JaniM) [inline bot.js]: Prime usecase for a `switch`.

Comment (JaniM) [reply bot.js]: Also are you sure `add` doesn't require authorization of any kind?

File: src/add.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    const platform = userMessageContent[platformIndex];
+    const channelName = userMessageContent.slice(platformIndex + 1).join(" ");
+    console.log(
+      `User: ${messageStack.member.user.tag} is adding ${channelName} to ${platform}`
+    );
+
+    if (platform === "youtube") {
+      addYoutubeChannelToJson(channelName, messageStack);
+    } else if (platform === "twitch") {
+      addTwitchChannelToJson(channelName, messageStack);
```
Comment (JaniM) [inline src/add.js]: We don't `await` the adds so the outer try doesn't catch errors in them.

File: bot.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    receiveCommandAdd(userMessageContent, messageStack);
+  }
+
+  if (commandSentByUser === "remove") {
```
Comment (JaniM) [inline bot.js]: Same for `remove`

File: bot.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  fs.readFileSync(path.join(process.cwd(), "database", "youtube.json"), "utf8")
+);
+const twitchJson = JSON.parse(
+  fs.readFileSync(path.join(process.cwd(), "database", "twitch.json"), "utf8")
+);
+
+const guilds = channelJson.guilds;
+
+let twitchJsonChannels = twitchJson.twitch;
+let youtubeJsonChannels = youtubeJson.youtube;
```
Comment (JaniM) [inline bot.js]: These aren't even used.

File: bot.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+let scheduledStartTime = [];
+let guildsNotificationChannels = [];
+let guildsTsumegoChannels = [];
+let guildsPvPChannels = [];
```
Comment (JaniM) [inline bot.js]: I highly advice trying to structure the program such that you don't need global mutable state.
Most of the time taken during my review was trying to figure out what these do.

Comment (JaniM) [reply bot.js]: Also `guildsTsumegoChannels` and `guildsPvPChannels` aren't used.

File: src/live.js (hunk 2 above)
Comment (JJLDonley) [reply src/live.js:17]: This shouldn't be the case. It should push into the global array. However I also was not correctly sorting and filtering live videos. I changed .include to (scheduledStartTime.some(video => video.URL === videoScheduled.URL))

File: src/live.js (hunk 3 above)
Comment (JJLDonley) [reply src/live.js:77]: Yeah this was not the intended effect, and instead I opted to remove last messages and instead just search to the posted channel and post to it directly

```js
async function postLiveVideosToChannels(
  liveStreamVideos,
  notificationChannels,
  client
) {
  for (const liveStreamVideo of liveStreamVideos) {
    const message = `${liveStreamVideo.channelName} is live!\n${liveStreamVideo.videoTitle}\n${liveStreamVideo.videoURL}`;
    for (const channel of notificationChannels) {
      const messages = await client.channels.cache
        .get(channel)
        .messages.fetch({ limit: 25 });
      const messageWithUrl = messages.find((msg) =>
        msg.content.includes(liveStreamVideo.videoURL)
      );
      if (!messageWithUrl) {
        client.channels.cache.get(channel).send(message);
        console.log(`Posted ${message} to ${channel}`);
      }
    }
  }
}
```

**Pull request closed without merging**